estimators/all_estimators.py

```diff
@@ -0,0 +1,279 @@
+    var = mse_control / n0 + mse_treated / n1
+    tau_estimated = (y1_c - y0_c).mean()
+
+    bias = [e_control, e_treated]
+
+    metric = [mse_control, mse_treated]
+
+    return tau_estimated, metric, bias, var
+
+
+def estimator_template(data, param_method, seed=0):
+    """Template to add new estimators.
+
+    This function normally returns four values. Some estimators do not return these values.
+    In this case, you can either calculate manually (adding lines of code) or return None.
+    The only exception is to the tau_estimated (IT SHOULD ALWAYS BE RETURNED).
+
+    IMPORTANT: the prefix 'estimator_' should be kept in order to avoid NotImplementedError errors.
+
+    :param data:
+    :param param_method:
+    :param seed:
+    :return:
+    """
+
+    np.random.seed(seed)
+    logger.debug('Running YOUR METHOD')
+
+    # 1) Call your estimator.
+    # 2) Train/Estimate tau_estimated using data and param_method.
+
+    tau_estimated = 999
+    metric = [None, None]
+    bias = [None, None]
+    var = None
+
+    return tau_estimated, metric, bias, var
+
+
+def _fit_base_models(model, data, data_all, param_method):
+    """Predicts the outcome on the full data.
+
+    :param data: tf.data.Dataset (control or treated).
+    :param model: tf.keras.Model.
+    :param dataset_all: tf.data.Dataset (all, for prediction).
+
+    :return: y_pred: [], predictions on data_all.
+     mse: array with mse on the control and treated group.
+     bias: array with bias on the control and treated group.
+     t: [], treatment assigment on data_all.
+     y_obs: [], observed outcome on data_all.
+    """
+
+    history = model.fit(data, steps_per_epoch=param_method['steps'], epochs=param_method['epochs'], verbose=2)
+    try:
+        mse = history.history['mean_squared_error'][-1]
+    except KeyError:
+        mse = history.history['mse'][-1]
+
+    bias = history.history['mae'][-1]
+    logger.debug('Calculate Predictions')
+    y_pred, y_obs, t = _predict_base_models(data=data_all, model=model, steps=param_method['steps'])
+    logger.debug('Done Predictions')
+    return y_pred, mse, bias, t, y_obs
+
+
+def _predict_base_models(data, model, steps):
+    """Predicts the outcome on the full data.
+
+    :param data: tf.data.Dataset.
+    :param model: tf.keras.Model.
+    :return: arrays with predicted outcome, observed outcome, and treat. assignment.
+    """
+    y_pred = []
+    y_obs = []
+    t = []
+    for i, (batch_x, batch_y, batch_t) in enumerate(data):
+        y_pred.append(model.predict_on_batch(batch_x))
+        y_obs.append(batch_y.numpy())
+        t.append(batch_t.numpy())
+        if i > steps:
+            break
+
+    y_pred_flat = np.concatenate(y_pred).ravel()
+    y_obs_flat = np.concatenate(y_obs).ravel()
+    t_flat = np.concatenate(t).ravel()
+
+    return np.array(y_pred_flat), np.array(y_obs_flat), np.array(t_flat)
+
+
+def _calculate_propensity_score(param_method, data, y_treated_hat, y_control_hat, t, y):
```

Comment:
Why is a function `_calculate_propensity_score` computing the treatment effect?

This is probably best to be separated into it's own function.

Reply:
That makes sense. I moved the computing of the treatment effect out of the function `_calculate_propensity_score`